### ISSUE TITLE
Error when applying CosMx data using squidpy library as input

### ISSUE THREAD

**Issue by ayumatsubo** [dpeerlab/ENVI#1]
Thank you for creating and sharing the very useful tool with the community. I tried to apply CosMx data to this tool, as demonstrated in the tutorial (https://github.com/dpeerlab/ENVI/blob/main/MOp_MERFISH_tutorial.ipynb), but I encountered an error. I loaded the CosMx data using the squidpy library as input (https://squidpy.readthedocs.io/en/stable/external_tutorials/tutorial_nanostring.html). The error occurs at the line `ENVI_Model = ENVI.ENVI(spatial_data = st_data, sc_data = sc_data)`. Could you please provide some guidance on how to resolve this error? Also, I am wondering if this method may not be applicable to CosMx. I would greatly appreciate any help you can offer. 

Thank you.

Error is following:
![スクリーンショット 2023-04-24 19 06 57](https://user-images.githubusercontent.com/26326449/233967749-cbf21ef0-e4e3-4cde-8eeb-35de4412134e.png)

**Comment by DoronHav** [comment]
Thanks for being interested in our method! 

This bug originates from the data (spatial_data.X) being in sparse format, which we currently do not support. Transform the data into dense and it should work!

**Comment by ayumatsubo** [comment]
Thank you for your quick response.

Using the dense matrix, it worked and no errors occurred. However, it seems that the output of ENVI resulted in all "nan" values for the loss and all other outputs. Could this issue be specific to our data? I loaded the data using `squidpy` and transformed it using the `anndata.X.toarray()` function.

Thank you.

![スクリーンショット 2023-04-27 18 32 44](https://user-images.githubusercontent.com/26326449/234822424-5500e5a9-cbb9-49b6-8223-364a7e3ff150.png)

**Comment by DoronHav** [comment]
This can occasionally happen. Most of the time, this can be easily fixed by lowering the learning rate in the ENVI_Model.Train function (set LR=0.00001 instead of the default LR = 0.0001).  

nans can also be caused from extremely high count values (in the thousands) in either the spatial of scRNA-seq datasets, or from genes with very little expression, both leading to instabilities in the deep probabilistic modeling. We recommend removing lowly expressed genes and thresholding high values if errors persist. 

We would love to get more feedback from users of our method. Feel free email me at doron.haviv12@gmail.com if you have any more comments.
